Apollo/Devices/Fade.cs

```diff
@@ -115,27 +147,45 @@ class FadeInfo {
 
         void Generate() => Generate(Preferences.FadeSmoothness);
 
-        void Generate(double smoothness) {
+        void Generate(double smoothness)
+        {
             if (_colors.Count < 2 || _positions.Count < 2) return;
+            if (_types.Count < _colors.Count - 1) return;
 
             List<Color> _steps = new List<Color>();
             List<int> _counts = new List<int>();
-            List<int> _cutoffs = new List<int>() {0};
+            List<int> _cutoffs = new List<int>() { 0 };
+
 
-            for (int i = 0; i < _colors.Count - 1; i++) {
+            for (int i = 0; i < _colors.Count - 1; i++)
+            {
                 int max = new int[] {
                     Math.Abs(_colors[i].Red - _colors[i + 1].Red),
                     Math.Abs(_colors[i].Green - _colors[i + 1].Green),
                     Math.Abs(_colors[i].Blue - _colors[i + 1].Blue),
                     1
                 }.Max();
 
-                for (int k = 0; k < max; k++) {
-                    _steps.Add(new Color(
-                        (byte)(_colors[i].Red + (_colors[i + 1].Red - _colors[i].Red) * k / max),
-                        (byte)(_colors[i].Green + (_colors[i + 1].Green - _colors[i].Green) * k / max),
-                        (byte)(_colors[i].Blue + (_colors[i + 1].Blue - _colors[i].Blue) * k / max)
-                    ));
+                switch (_types[i])
```

Comment:
This switch is sorta ugly to see repeating here, how about something like [this](https://github.com/mat1jaczyyy/apollo-studio/blob/f3a7e1a8b945250579eb59ce6e4264519944cea3/Apollo/Elements/Launchpad.cs#L115) except instead of `byte[]` you have lambdas that do the proper thing? so the right one is autocalled. should be `Func<arg1, arg2, ... argn,  ret>`